### PR TITLE
Eliminate per-shard heap allocations with InlineVec and upfront capacity planning

### PR DESCRIPTION
**Pull request by ahrav**
## Summary

Reduces heap allocation pressure in the coordination hot path by introducing `InlineVec<T, N>` — a stack-first small vector that stores up to N elements inline before spilling to a heap `Vec`. The primary integration replaces `Vec<ShardId>` with `InlineVec<ShardId, 8>` for shard spawn tracking, eliminating one allocation per shard in the 99%+ common case (0–2 children). Additionally, coordinator `HashMap`s are pre-sized with initial capacity, `BTreeMap` is replaced with `AHashMap` for cooldowns, and a `CoordinatorConfig` memory budget estimator is added for deployment planning.

- **21 files changed** (+1,621 / −70)
- **Risk**: Medium — new `unsafe` code in `InlineVec`, backed by Miri, Kani (9 proofs), property tests, and fuzz harness

### New: `InlineVec<T, N>` (`gossip-stdx`)

Stack-first small vector backed by `[MaybeUninit<T>; N]`. One-way spill to heap on overflow. Full API: `new`, `push`, `extend_from_slice`, `as_slice`, `iter`, `from_slice`, `From<Vec>`, `FromIterator`, `Clone`, `Drop`, `PartialEq/Eq`, `Debug`, `Default`, `IntoIterator`.

Verification layers:

| Layer | What it catches |
|-------|-----------------|
| 30+ unit tests | Logic errors, edge cases, ZST, N=1 |
| Drop tracking | Double-free, leak |
| Clone-panic test | Unwind safety |
| Property tests (proptest, 256×200 ops) | Behavioral divergence from Vec oracle |
| Fuzz target (libfuzzer) | Crash/UB under random input |
| Kani proofs (9) | Out-of-bounds, uninitialized reads (every `unsafe` block) |
| Miri | UB, use-after-free, stacked borrows |

Criterion benchmarks added for push, from_slice, extend, spill, as_slice, and Vec baseline.

### Coordinator integration (`gossip-contracts`)

- `ShardRecord.spawned`: `Vec<ShardId>` → `SpawnedList` (`InlineVec<ShardId, 8>`)
- `claim_cooldowns`: `BTreeMap` → `AHashMap` (O(1) vs O(log N) for point lookups; deterministic debug output wasn't worth the cost)
- All `HashMap`/`HashSet` constructors pre-sized with `with_capacity_and_hasher`
- New `CoordinatorConfig` with `dev_defaults()`, `staging_defaults()`, `prod_defaults()` presets and `memory_budget()` estimator

### Size limit tightening (data-justified)

- `MAX_TOKEN_SIZE`: 64 KiB → 16 KiB (largest observed token ~15 KB)
- `MAX_METADATA_SIZE`: 64 KiB → 16 KiB (largest observed metadata ~8 KB)
- `AcquireResult` size assertion: ≤224 → ≤288 bytes (inline storage is larger)

### CI

- Kani verification job (nightly/workflow_dispatch only)

## Why

1. **Allocation elimination**: 99%+ of shards have 0–2 spawned children. `Vec` heap-allocates on first push; `InlineVec<_, 8>` keeps them on the stack.
2. **Upfront capacity**: `with_capacity_and_hasher` prevents resize-copy cascades during coordinator warmup.
3. **BTreeMap → AHashMap**: Point lookups only — O(log N) wasn't justified.
4. **Tighter limits**: 64 KiB was a placeholder. Real data fits in 16 KiB with margin.

## Unsafe code audit

8 `unsafe` blocks in `InlineVec`, each with `// SAFETY:` comments. Critical design: `Repr` has **no `Drop` impl** — `InlineVec::drop` handles all cleanup, preventing double-free during the `mem::replace` spill path. All blocks covered by Kani proofs + Miri.

## Test plan

- [ ] `cargo test --all` passes
- [ ] `cargo +nightly miri test -p gossip-stdx` passes
- [ ] `cargo bench -p gossip-stdx --bench inline_vec` runs cleanly
- [ ] Kani proofs pass (CI nightly job)
- [ ] Existing simulation invariant tests pass with InlineVec integration

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/gossip-contracts/src/coordination/in_memory.rs [hunk 1]
```diff
@@ -112,6 +112,129 @@ use gossip_stdx::RingBuffer;
...
+    ///
+    /// Where:
+    /// - `S` = `max_total_shards`, `B_s` = `per_shard_budget`
+    /// - `R` = `max_runs`, `B_r` = `per_run_budget`
+    /// - `W` = `max_workers`
+    /// - 736 = fixed-size portion of `ShardRecord`
+    /// - 72 = per-entry HashMap overhead (key + bucket metadata)
+    /// - 576 = fixed-size portion of `RunRecord`
+    /// - 80 = per-entry HashMap overhead for run entries
+    /// - 16 = per-entry in claim_cooldowns
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-contracts/src/coordination/in_memory.rs]: Good catch — both constants were already stale:
- `ShardRecord` is **776** bytes (was hardcoded as 736)
- `RunRecord` is **512** bytes (was hardcoded as 576)

Updated the constants in `memory_budget()` and added a `memory_budget_constants_match_struct_sizes` test that asserts `size_of::<ShardRecord>()` and `size_of::<RunRecord>()` match the hardcoded values. Future drift will be caught immediately by CI.

File: crates/gossip-stdx/src/inline_vec.rs [hunk 2]
```diff
@@ -0,0 +1,1094 @@
...
+            Repr::Inline { buf, len } if (*len as usize) + slice.len() <= N => {
+                let start = *len as usize;
+                for (i, item) in slice.iter().enumerate() {
+                    // SAFETY: start + i < start + slice.len() <= N, so
+                    // buf[start + i] is in bounds. The slot is uninitialized
+                    // (past the last initialized slot). We write a valid T
+                    // and update len after the loop to maintain the invariant.
+                    unsafe { buf.get_unchecked_mut(start + i).write(item.clone()) };
+                }
+                *len += slice.len() as u32;
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-stdx/src/inline_vec.rs]: Confirmed — same class of bug as the Clone issue. Added a test (`extend_from_slice_panic_drops_partial_clones`) that demonstrates the leak.

Fixed by incrementing `len` after each successful clone write instead of batching the update after the loop. This ensures `InlineVec::drop` accounts for all partially-written elements if a subsequent `T::clone()` panics.
